Convert core toolbar buttons into ToolbarButton

diegohaz opened the pull request:
## Context

This PR is part of #18619, whose main goal is to implement [roving tabindex](https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex) on the `@wordpress/components`' `Toolbar` component and to use it on the block toolbars so they become a single tab stop as recommended by the [WAI-ARIA Toolbar Pattern](https://www.w3.org/TR/wai-aria-practices/#toolbar). Related issues are #15331 and #3383.

![GIF showing the navigation through the Gutenberg UI using the tab key](https://user-images.githubusercontent.com/3068563/80681665-825c4000-8a97-11ea-9bd9-0ac570b82591.gif)

This PR improves [`NavigableToolbar`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/block-editor/src/components/navigable-toolbar/index.js) and converts the core toolbar buttons (those buttons that appear on all blocks: movers, block switcher and more options) into [`ToolbarButton`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/components/src/toolbar-button/index.js) or [`ToolbarItem`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/components/src/toolbar-item/index.js).

*`ToolbarItem` is the generic headless component that is used by `ToolbarButton` underneath and can be used to make other components, such as [`DropdownMenu`](https://github.com/WordPress/gutenberg/blob/fd5816e2fed2f241c4eeaea99d94b53af14293db/packages/components/src/dropdown-menu/index.js), aware of the context state provided by the parent [`Toolbar`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/components/src/toolbar/toolbar-container.js) component, which enables the roving tabindex method.*

This also converts a few other buttons so the block toolbar on rich text blocks and other basic blocks already works with roving tabindex.

## `Toolbar` vs. `ToolbarGroup`

> Or `.components-toolbar` vs. `.components-toolbar-group`

Most of the changes on this PR are related to this. On #18534, `Toolbar` was renamed to [`ToolbarGroup`](https://github.com/WordPress/gutenberg/blob/e5dac46fcc0fb44d59d47182a28348a056ab431b/packages/components/src/toolbar-group/index.js), but the old class name (`components-toolbar`) was kept as there's a lot of code relying on it.

However, when `ToolbarGroup` is used within the accessible `Toolbar` component, the class name will be `components-toolbar-group`. That's why it appears duplicated in many parts.

## `NavigableToolbar`

`NavigableToolbar` is the component that is currently used by block toolbars and by the top toolbar. It handles focus when the user presses <kbd>Alt</kbd>+<kbd>F10</kbd>.

With this PR, if all the tabbable DOM elements inside `NavigableToolbar` are rendered by `ToolbarButton` (or `ToolbarItem`), then `NavigableToolbar` will render `Toolbar` instead of [`NavigableMenu`](https://github.com/WordPress/gutenberg/blob/fd5816e2fed2f241c4eeaea99d94b53af14293db/packages/components/src/navigable-container/menu.js). If there's any other tabbable element, it'll work just like it does right now on master.

https://github.com/WordPress/gutenberg/blob/70937da71e8a44a71a82e7814e6eefe7b6104c8e/packages/block-editor/src/components/navigable-toolbar/index.js#L80-L101

This is so we can gradually migrate to this approach, and not break blocks that aren't using `ToolbarButton` for their toolbar buttons, like plugin custom blocks or other blocks like Image (when there's an image). They still work with the <kbd>Tab</kbd> key.

![GIF showing the navigation through the Gutenberg UI using the tab key](https://user-images.githubusercontent.com/3068563/80684954-dae20c00-8a9c-11ea-86f0-d85882f07993.gif)

In short, this code would render `NavigableMenu` because, even though the first item is a `ToolbarButton`, the others are `Button`, `button` and `DropdownMenu`, which aren't aware of `ToolbarContext` and, therefore, can't join the roving tabindex party:

```jsx
<NavigableToolbar aria-label="Block contextual toolbar">
  <ToolbarButton />
  <Button />
  <button />
  <DropdownMenu />
</NavigableToolbar>
```

The code below would render `Toolbar` with roving tabindex enabled:

```jsx
<NavigableToolbar aria-label="Block contextual toolbar">
  <ToolbarButton />
  <ToolbarItem>
    { ( itemProps ) => ( <Button { ...itemProps } /> ) }
  </ToolbarItem>
  <ToolbarItem>
    { ( itemProps ) => ( <button { ...itemProps } /> ) }
  </ToolbarItem>
  <ToolbarItem>
    { ( itemProps ) => ( <DropdownMenu toggleProps={ itemProps } /> ) }
  </ToolbarItem>
</NavigableToolbar>
```

## How to test

- Check rich text blocks and other basic blocks are working with roving tabindex.
- Check if blocks that contain non-`ToolbarItem` buttons are working just like today on master.
- Check if block toolbars are correctly styled, including nested blocks with horizontal movers.
- The top toolbar should work just like on master as the other buttons there haven't been converted yet.

## What's next

- Convert the Top Toolbar buttons.
- Convert toolbar buttons on other blocks.

diegohaz committed:
SlotFill initial implementation

diegohaz committed:
Add manifest-devhub.json

diegohaz committed:
Accept as prop on Slot

diegohaz committed:
Update stories

diegohaz committed:
Update README.md

diegohaz committed:
Update code

diegohaz committed:
Add slot-fill2 entries to components index file

diegohaz committed:
Add unit tests

diegohaz committed:
Merge branch 'master' into update/slot-fill

diegohaz committed:
Fix git conflicts

diegohaz committed:
Merge branch 'master' into update/slot-fill

diegohaz committed:
Convert fixed toolbar buttons into ToolbarButton

diegohaz committed:
Lint code

diegohaz committed:
Merge branch 'update/slot-fill' into update/fixed-toolbar-buttons

diegohaz committed:
Merge branch 'master' into update/fixed-toolbar-buttons

diegohaz committed:
Update code

diegohaz committed:
Adjustments

diegohaz committed:
Temporarily fix toolbar buttons styles

diegohaz committed:
Merge branch 'master' into update/fixed-toolbar-buttons

diegohaz committed:
Revert block toolbar styles

diegohaz committed:
Fix styles

diegohaz committed:
Lint code

diegohaz committed:
Update e2e-tests

diegohaz committed:
Update styles

diegohaz committed:
Merge branch 'master' into update/fixed-toolbar-buttons

diegohaz committed:
Merge branch 'master' into update/fixed-toolbar-buttons

diegohaz committed:
Update styles

diegohaz committed:
Fix unit tests

diegohaz committed:
Fix some e2e tests

diegohaz committed:
Update NavigableToolbar

diegohaz committed:
Attempt to fix e2e tests

diegohaz committed:
Revert Attempt to fix e2e tests

diegohaz committed:
Fix navigable-toolbar tests

diegohaz committed:
Upgrade reakit

diegohaz committed:
Fix e2e tests

diegohaz committed:
Update styles

diegohaz committed:
Update ToolbarGroup styles

diegohaz committed:
Update NavigableToolbar

diegohaz committed:
Refactor Block Controlls Fill/Slot code

diegohaz committed:
Update NavigableToolbar

diegohaz committed:
Update styles

diegohaz committed:
Add comment on block-toolbar/utils.js

diegohaz committed:
Refactor format-toolbar code

diegohaz commented:
This is ready for review! 🙂

gziolo commented:
I read the description and it all is explained very well. I understood why the fallback is necessary and it's handled nicely. The only question is how can we encourage plugin developers to migrate to the new approach so we avoid mixes experience where tabbing behaves differently depending on the underlying implementation? In the default WP installation, it will consistently use the new behavior. However, for 3rd party blocks, it will most likely defer to the legacy behavior.

diegohaz commented:
What I'm thinking is: once it's internally stable, since we can identify when there are non-`ToolbarItem` tabbable elements being used within the block toolbars, we can emit a deprecation warning saying that `ToolbarButton` (or `ToolbarItem`) should be used instead.

But I'm not sure if this will be enough, if third party block authors pay attention to those warnings.

ItsJonQ commented on packages/block-editor/src/components/navigable-toolbar/index.js:
```diff
@@ -1,28 +1,93 @@
  */
-import { NavigableMenu } from '@wordpress/components';
-import { useRef, useEffect, useCallback } from '@wordpress/element';
+import { NavigableMenu, Toolbar } from '@wordpress/components';
```
@diegohaz It looks like this is where all the magic happens ✨ 😊

ItsJonQ commented on packages/block-editor/src/components/block-toolbar/utils.js:
```diff
@@ -153,6 +158,8 @@ export function useShowMoversGestures( {
 	};
 }
 
+let requestAnimationFrameId;
```
Thank you for improving this! 🙏

gziolo replied:
It could use some explanation in the code. There can be multiple instances of the block toolbar. I read that it's related to the focus that can be in one place so it's perfectly fine. Anyway, it's something that might be unclear for the reader at first glance.

ItsJonQ reviewed (approved):
🚀 from me!! Amazing work on this @diegohaz !

I tested this locally, and the tab sequencing works as expected and advertised 😊 .
I reviewed the code as well, and everything looks great!
It looks like most of the code involved renaming of `Toolbar` to `ToolbarGroup` (which makes sense, and you had already mentioned it)

I focused a lot on going through `navigable-toolbar`, as it looks like a lot of new focus based mechanics were added there. Wonderful stuff! I learned a thing or two as well. It's nice that the focus/interaction complexity is abstracted away and contained within that component.

Looking forward to seeing how Toolbar evolves from here!

Again, thank you for all of your work! Both on this PR as well as on all the Reakit internals that powers these components ❤️

diegohaz committed:
Fix styles on horizontal movers

diegohaz committed:
Update trailing dropdown toolbar button styles

diegohaz committed:
Use MutationObserver instead of relying on the next repaint

diegohaz committed:
Fix drag and drop

diegohaz committed:
useUpdateLayoutEffect for MutationObserver

diegohaz committed:
Update NavigableToolbar code

diegohaz committed:
Update NavigableToolbar code

diegohaz commented:
Fixed a few issues:

- [x] The horizontal movers weren't being displayed correctly (5bb5f499fbf990d6ceedc53043e93f0b4c96ca2c)
- [x] When the last item in a toolbar group is a dropdown (like the "More rich text controls" button), it was getting an additional `padding-left`. (9b600751f9c18a2da371ae415eccda23c838df7b)
- [x] Drag and drop wasn't working (ed633cc20cf79c54df5e03c2817ed8b0cf3d8270)

And this one is really interesting:

1. Add an image block.
2. Upload an image.
3. Press <kbd>Shift</kbd>+<kbd>Tab</kbd> until you reach the toolbar.
4. Continue pressing <kbd>Shift</kbd>+<kbd>Tab</kbd> and/or arrow keys to see the bug.

![May-01-2020 00-33-25](https://user-images.githubusercontent.com/3068563/80779902-71690880-8b43-11ea-89e3-5e65cc368673.gif)

This was happening because some non-`ToolbarItem` elements are rendered asynchronously. To mitigate that, I was also checking the next browser repaint:

https://github.com/WordPress/gutenberg/blob/9b600751f9c18a2da371ae415eccda23c838df7b/packages/block-editor/src/components/navigable-toolbar/index.js#L44-L49

But those toolbar buttons can be added to the toolbar much later, at any time. On the Image block, for example, they're added after the user has selected the image to upload.

Fixed it by using [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) instead:

https://github.com/WordPress/gutenberg/blob/2399dbd272a37bc329798e9d27989ba4d7e73348/packages/block-editor/src/components/navigable-toolbar/index.js#L49-L59

gziolo commented:
@diegohaz, can you confirm the overall strategy for the whole process:

- `<Toolbar __experimentalAccessibilityLabel="Block toolbar" />` changes the toolbar to behave according to WAI-ARIA specs (fixes also this issue from a11y audit: https://github.com/WordPress/gutenberg/issues/15331), otherwise it’s just a proxy to `ToolbarGroup` to make it backward compatible
- `<ToolbarGroup />` is used for grouping items, it also turns into dropdown when the flag for collapsing is enabled
- `<ToolbarItem />` and `<ToolbarButton />` need to be used for individual items, or the whole toolbar will fallback to the incorrect legacy behavior for toolbars

I want to make sure everyone can follow everything properly and be able to review follow-up PRs, including those where we updated docs to reflect new recommendations.

diegohaz commented:
@gziolo

> - `<Toolbar __experimentalAccessibilityLabel="Block toolbar" />` changes the toolbar to behave according to WAI-ARIA specs (fixes also this issue from a11y audit: https://github.com/WordPress/gutenberg/issues/15331), otherwise it’s just a proxy to `ToolbarGroup` to make it backward compatible
> - `<ToolbarGroup />` is used for grouping items, it also turns into dropdown when the flag for collapsing is enabled
> - `<ToolbarItem />` and `<ToolbarButton />` need to be used for individual items, or the whole toolbar will fallback to the incorrect legacy behavior for toolbars

Exactly that! 👍

I'm going to add a comment on Toolbar to avoid confusion for people looking at the code before the old behavior is deprecated.

diegohaz committed:
Add comment on Toolbar

diegohaz committed:
Move bubblesVirtually prop so React Native still works

enriquesanchez commented:
Hi @diegohaz!

To clarify, this is not yet implemented in all blocks, correct? I'm seeing weird behavior with blocks such as columns and group.

diegohaz commented:
@enriquesanchez That's correct! It should work for columns though. Could you elaborate on how it’s weird? I just did a quick test and didn’t find the problem.

gziolo commented on packages/block-editor/src/components/block-controls/index.js:
```diff
@@ -10,15 +20,32 @@ import { ifBlockEditSelected } from '../block-edit/context';
+				// So we re-create the Provider in this subtree.
+				const value = ! isEmpty( fillProps ) ? fillProps : null;
+				return (
+					<ToolbarContext.Provider value={ value }>
```
I remember having to do the same trick in the past for other fills. It's a bit unfortunate that we need this trick but as long it works it feels like a good trade-off.

gziolo commented on packages/block-editor/src/components/navigable-toolbar/index.js:
```diff
@@ -1,28 +1,103 @@
+	if ( isAccessibleToolbar ) {
+		return (
+			<Toolbar
+				__experimentalAccessibilityLabel={ props[ 'aria-label' ] }
```
Once it becomes an official prop, we should refactor code to stop using `aria-label` directly and convert it to pass-through with `props`.

gziolo commented on packages/components/src/toolbar-button/index.js:
```diff
@@ -21,6 +21,9 @@ function ToolbarButton( {
 	extraProps,
 	children,
+	title,
+	isActive,
```
Aside: So much legacy here :( Ideally, we should be promoting `label` and `isPressed` to make it easier for developers to use API. At the same time, `disabled` in the `Button` should be really `isDisabled` 😃 
Nothing actionable for now, I just wanted to complain a bit 🙃

diegohaz replied:
I think we should eventually deprecate those props and make the API the same as on `Button`. I chose to add those props here because there is already a ton of code using `ToolbarButton` with them, and the migration would be more painful.

gziolo replied:
Agreed 💯

gziolo reviewed (approved):
Awesome work Diego. In my testing, only Block Toolbar uses new behavior when it isn't docked in the header. I'm looking forward to seeing the rest of the toolbars converted 🙌 

There was the following comment from _alexstine_ on WordPress Slack in #accessibility channel:

> Testing in Firefox with NV Access.
> 1. Can tab in to the toolbar.
> 2. Can use arrow navigation to move through the toolbar.
> 3. Tab will move to each toolbar element.

(3) seems concerning, but I should be more clear and let everyone know that this new behavior applies only to the floating toolbar.

> My final recommendation would be to replace the div that wraps the toolbar with a section element. This should allow you to jump by landmark to the Document tools section and then here it announced in some screen readers when you leave that section and land on the "Publish" button.
> I've not mocked up the code so I cannot say how it would work exactly, but I think it would be worth a try to add a bit more support from jumping to and from that Document tools section.

Sharing also the above for visibility, I don't think it's something we should include in this PR.

diegohaz committed:
Merge branch 'master' into update/fixed-toolbar-buttons

diegohaz committed:
Remove missing comment from styles

enriquesanchez commented:
@diegohaz Oh, I apologize! There was confusion on my part. I just tested again and works as expected.

diegohaz commented:
@enriquesanchez No worries! Thanks for testing this! ❤️

SergioEstevao reviewed (commented):
Just circling back to this PR, while we appreciate the change that was done, if a `.native` file is being deleted it's a good indicator that this code is being used in GB-mobile too. 

It would be nice in the future to ping the GB-mobile team just to make sure that the change is ok for the GB-mobile side of things.

diegohaz commented:
@SergioEstevao Thanks for the heads up! Created #22229 to re-add that file.
